File: app/src/main/java/com/elementary/tasks/reminder/preview/ReminderPreviewActivity.java
package com.elementary.tasks.reminder.preview;

import android.app.AlarmManager;
import android.app.AlertDialog;
import android.content.Intent;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import com.elementary.tasks.R;
import com.elementary.tasks.core.ThemedActivity;
import com.elementary.tasks.core.data.models.Place;
import com.elementary.tasks.core.data.models.Reminder;
import com.elementary.tasks.core.fragments.AdvancedMapFragment;
import com.elementary.tasks.core.interfaces.MapCallback;
import com.elementary.tasks.core.utils.Constants;
import com.elementary.tasks.core.utils.Dialogues;
import com.elementary.tasks.core.utils.IntervalUtil;
import com.elementary.tasks.core.utils.LogUtil;
import com.elementary.tasks.core.utils.Module;
import com.elementary.tasks.core.utils.Prefs;
import com.elementary.tasks.core.utils.RealmDb;
import com.elementary.tasks.core.utils.ReminderUtils;
import com.elementary.tasks.core.utils.Sound;
import com.elementary.tasks.core.utils.TimeUtil;
import com.elementary.tasks.core.view_models.reminders.ReminderViewModel;
import com.elementary.tasks.reminder.create_edit.CreateReminderActivity;
import com.elementary.tasks.databinding.ActivityReminderPreviewBinding;
import com.elementary.tasks.databinding.ListItemTaskBinding;
import com.elementary.tasks.databinding.NoteListItemBinding;
import com.elementary.tasks.google_tasks.TaskActivity;
import com.elementary.tasks.google_tasks.TaskItem;
import com.elementary.tasks.google_tasks.TasksConstants;
import com.elementary.tasks.notes.NoteItem;
import com.elementary.tasks.notes.NotePreviewActivity;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;

import java.io.File;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProviders;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class ReminderPreviewActivity extends ThemedActivity {

    private static final String TAG = "ReminderPreviewActivity";

    private ActivityReminderPreviewBinding binding;
    private AdvancedMapFragment mGoogleMap;
    private ReminderViewModel viewModel;

    @NonNull
    private List<Long> list = new ArrayList<>();
    @Nullable
    private NoteItem mNoteItem;
    @Nullable
    private TaskItem mTaskItem;
    @NonNull
    private Handler mUiHandler = new Handler(Looper.getMainLooper());
    @Nullable
    private Reminder reminder;

    @NonNull
    private ReadyListener mReadyCallback = object -> {
        if (object == null) return;
        if (object instanceof NoteItem) {
            this.mNoteItem = (NoteItem) object;
            showNote();
        } else if (object instanceof TaskItem) {
            this.mTaskItem = (TaskItem) object;
            showTask();
        }
    };
    @NonNull
    private MapCallback mMapReadyCallback = new MapCallback() {
        @Override
        public void onMapReady() {
            mGoogleMap.setSearchEnabled(false);
            if (reminder != null) showMapData(reminder);
        }
    };
    @NonNull
    private GoogleMap.OnMarkerClickListener mOnMarkerClick = new GoogleMap.OnMarkerClickListener() {
        @Override
        public boolean onMarkerClick(Marker marker) {
            mGoogleMap.moveCamera(marker.getPosition(), 0, 0, 0, 0);
            return false;
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        int id = getIntent().getIntExtra(Constants.INTENT_ID, 0);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_reminder_preview);
        initActionBar();
        initViews();

        initViewModel(id);
    }

    private void initViewModel(int id) {
        ReminderViewModel.Factory factory = new ReminderViewModel.Factory(getApplication(), id);
        viewModel = ViewModelProviders.of(this, factory).get(ReminderViewModel.class);
        viewModel.reminder.observe(this, reminder -> {
            if (reminder != null) {
                showInfo(reminder);
            }
        });
        viewModel.result.observe(this, commands -> {
            if (commands != null) {
                switch (commands) {
                    case DELETED:
                        closeWindow();
                        break;
                }
            }
        });
    }

    private void showTask() {
        if (mTaskItem != null) {
            ListItemTaskBinding binding = ListItemTaskBinding.inflate(LayoutInflater.from(this));
            binding.setTaskItem(mTaskItem);
            binding.setClick(v -> startActivity(new Intent(ReminderPreviewActivity.this, TaskActivity.class)
                    .putExtra(Constants.INTENT_ID, mTaskItem.getTaskId())
                    .putExtra(TasksConstants.INTENT_ACTION, TasksConstants.EDIT)));
            this.binding.dataContainer.addView(binding.getRoot());
        }
    }

    private void showNote() {
        if (mNoteItem != null) {
            NoteListItemBinding binding = NoteListItemBinding.inflate(LayoutInflater.from(this));
            binding.setNoteItem(mNoteItem);
            binding.noteClick.setOnClickListener(v ->
                    startActivity(new Intent(ReminderPreviewActivity.this, NotePreviewActivity.class)
                    .putExtra(Constants.INTENT_ID, mNoteItem.getKey())));
            this.binding.dataContainer.addView(binding.getRoot());
        }
    }

    private void showMapData(Reminder reminder) {
        Place place = reminder.getPlaces().get(0);
        double lat = place.getLatitude();
        double lon = place.getLongitude();
        binding.mapContainer.setVisibility(View.VISIBLE);
        binding.location.setVisibility(View.VISIBLE);
        binding.location.setText(String.format(Locale.getDefault(), "%.5f %.5f (%d)", place.getLatitude(), place.getLongitude(), reminder.getPlaces().size()));
        mGoogleMap.addMarker(new LatLng(lat, lon), reminder.getSummary(), true, true, place.getRadius());
    }

    private void showInfo(Reminder reminder) {
        this.reminder = reminder;
        if (reminder != null) {
            binding.statusSwitch.setChecked(reminder.isActive());
            if (!reminder.isActive()) {
                binding.statusText.setText(R.string.disabled);
            } else {
                binding.statusText.setText(R.string.enabled4);
            }
            binding.windowTypeView.setText(getWindowType(reminder.getWindowType()));
            binding.taskText.setText(reminder.getSummary());
            binding.type.setText(ReminderUtils.getTypeString(this, reminder.getType()));
            binding.itemPhoto.setImageResource(getThemeUtil().getReminderIllustration(reminder.getType()));
            long due = TimeUtil.getDateTimeFromGmt(reminder.getEventTime());
            if (due > 0) {
                binding.time.setText(TimeUtil.getFullDateTime(due, getPrefs().is24HourFormatEnabled(), false));
                String repeatStr = IntervalUtil.getInterval(this, reminder.getRepeatInterval());
                if (Reminder.isBase(reminder.getType(), Reminder.BY_WEEK)) {
                    repeatStr = ReminderUtils.getRepeatString(this, reminder.getWeekdays());
                }
                if (repeatStr != null) {
                    binding.repeat.setText(repeatStr);
                } else {
                    binding.repeat.setVisibility(View.GONE);
                }
            } else {
                binding.time.setVisibility(View.GONE);
                binding.repeat.setVisibility(View.GONE);
            }
            if (Reminder.isGpsType(reminder.getType())) {
                initMap();
            } else {
                binding.location.setVisibility(View.GONE);
                binding.mapContainer.setVisibility(View.GONE);
            }
            String numberStr = reminder.getTarget();
            if (!TextUtils.isEmpty(numberStr)) {
                binding.number.setText(numberStr);
            } else {
                binding.number.setVisibility(View.GONE);
            }

            File file = null;
            if (!TextUtils.isEmpty(reminder.getMelodyPath())) {
                file = new File(reminder.getMelodyPath());
            } else {
                String path = getPrefs().getMelodyFile();
                if (path != null && !Sound.isDefaultMelody(path)) {
                    file = new File(path);
                } else {
                    Uri soundUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION);
                    if (soundUri != null && soundUri.getPath() != null) {
                        file = new File(soundUri.getPath());
                    }
                }
            }
            if (file != null) binding.melody.setText(file.getName());

            int catColor = 0;
            if (reminder.getGroup() != null) {
                binding.group.setText(reminder.getGroup().getTitle());
                catColor = reminder.getGroup().getColor();
            }
            int mColor = getThemeUtil().getColor(getThemeUtil().getCategoryColor(catColor));
            binding.appBar.setBackgroundColor(mColor);
            if (Module.isLollipop()) {
                getWindow().setStatusBarColor(getThemeUtil().getNoteDarkColor(catColor));
            }
            binding.dataContainer.removeAllViewsInLayout();
            showAttachment(reminder);
            new Thread(new NoteThread(mReadyCallback, reminder.getNoteId())).start();
            new Thread(new TaskThread(mReadyCallback, reminder.getUuId())).start();
        }
    }

    private String getWindowType(int reminderWType) {
        int windowType = Prefs.getInstance(this).getReminderType();
        boolean ignore = Prefs.getInstance(this).isIgnoreWindowType();
        if (!ignore) {
            windowType = reminderWType;
        }
        return windowType == 0 ? getString(R.string.full_screen) : getString(R.string.simple);
    }

    private void showAttachment(Reminder reminder) {
        if (reminder != null) {
            if (reminder.getAttachmentFile() != null) {
                File file = new File(reminder.getAttachmentFile());
                binding.attachmentView.setText(file.getName());
                binding.attachmentView.setVisibility(View.VISIBLE);
            } else {
                binding.attachmentView.setVisibility(View.GONE);
            }
        } else {
            binding.attachmentView.setVisibility(View.GONE);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_reminder_preview, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int ids = item.getItemId();
        switch (ids) {
            case R.id.action_delete:
                removeReminder();
                return true;
            case android.R.id.home:
                closeWindow();
                return true;
            case R.id.action_make_copy:
                makeCopy();
                return true;
            case R.id.action_edit:
                editReminder();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void editReminder() {
        if (reminder != null) {
            startActivity(new Intent(this, CreateReminderActivity.class).putExtra(Constants.INTENT_ID, reminder.getUniqueId()));
        }
    }

    private void removeReminder() {
        if (reminder != null) {
            viewModel.moveToTrash(reminder);
        }
    }

    private void makeCopy() {
        if (reminder != null) {
            int type = reminder.getType();
            if (!Reminder.isGpsType(type) && !Reminder.isSame(type, Reminder.BY_TIME)) {
                showDialog();
            }
        }
    }

    private void closeWindow() {
        if (Module.isLollipop()) {
            mUiHandler.post(this::finishAfterTransition);
        } else {
            finish();
        }
    }

    public void showDialog() {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        calendar.set(Calendar.HOUR_OF_DAY, 0);
        calendar.set(Calendar.MINUTE, 0);
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);
        int hour = 0;
        int minute = 0;
        list.clear();
        List<String> time = new ArrayList<>();
        boolean is24 = getPrefs().is24HourFormatEnabled();
        do {
            if (hour == 23 && minute == 30) {
                hour = -1;
            } else {
                long tmp = calendar.getTimeInMillis();
                hour = calendar.get(Calendar.HOUR_OF_DAY);
                minute = calendar.get(Calendar.MINUTE);
                list.add(tmp);
                time.add(TimeUtil.getTime(calendar.getTime(), is24));
                calendar.setTimeInMillis(tmp + AlarmManager.INTERVAL_HALF_HOUR);
            }
        } while (hour != -1);
        AlertDialog.Builder builder = Dialogues.getDialog(this);
        builder.setTitle(R.string.choose_time);
        builder.setItems(time.toArray(new String[0]), (dialog, which) -> {
            dialog.dismiss();
            saveCopy(which);
        });
        AlertDialog alert = builder.create();
        alert.show();
    }

    private void saveCopy(int which) {
        LogUtil.d(TAG, "saveCopy: " + which);
        if (reminder != null) {
            viewModel.copyReminder(reminder, list.get(which), reminder.getSummary() + " - cope");
        }
    }

    private void initViews() {
        binding.switchWrapper.setOnClickListener(v -> switchClick());
        binding.mapContainer.setVisibility(View.GONE);
    }

    private void switchClick() {
        if (reminder != null) {
            viewModel.toggleReminder(reminder);
        }
    }

    private void initMap() {
        mGoogleMap = AdvancedMapFragment.newInstance(false, false, false, false, getPrefs().getMarkerStyle(), getThemeUtil().isDark());
        mGoogleMap.setCallback(mMapReadyCallback);
        mGoogleMap.setOnMarkerClick(mOnMarkerClick);
        getSupportFragmentManager().beginTransaction()
                .replace(binding.mapContainer.getId(), mGoogleMap)
                .addToBackStack(null)
                .commit();
    }

    private void initActionBar() {
        setSupportActionBar(binding.toolbar);
        getSupportActionBar().setDisplayShowTitleEnabled(false);
    }

    @Override
    public void onBackPressed() {
        finish();
    }

    private class NoteThread implements Runnable {

        private ReadyListener listener;
        private String uuId;

        NoteThread(ReadyListener listener, String uuId) {
            this.listener = listener;
            this.uuId = uuId;
        }

        @Override
        public void run() {
            NoteItem item = RealmDb.getInstance().getNote(uuId);
            runOnUiThread(() -> {
                if (listener != null && item != null) listener.onReady(item);
            });
        }
    }

    private class TaskThread implements Runnable {

        private ReadyListener listener;
        private String uuId;

        TaskThread(ReadyListener listener, String uuId) {
            this.listener = listener;
            this.uuId = uuId;
        }

        @Override
        public void run() {
            TaskItem item = RealmDb.getInstance().getTaskByReminder(uuId);
            runOnUiThread(() -> {
                if (listener != null && item != null) listener.onReady(item);
            });
        }
    }

    interface ReadyListener {
        void onReady(Object object);
    }
}

File: app/src/main/java/com/elementary/tasks/reminder/preview/ReminderPreviewActivity.kt
package com.elementary.tasks.reminder.preview

import android.app.ActivityOptions
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.recyclerview.widget.LinearLayoutManager
import com.elementary.tasks.R
import com.elementary.tasks.core.data.Commands
import com.elementary.tasks.core.data.ui.note.UiNoteList
import com.elementary.tasks.core.utils.TelephonyUtil
import com.github.naz013.common.datetime.DateTimeManager
import com.elementary.tasks.core.utils.params.Prefs
import com.github.naz013.ui.common.Dialogues
import com.elementary.tasks.databinding.ActivityReminderPreviewBinding
import com.elementary.tasks.googletasks.task.GoogleTaskActivity
import com.elementary.tasks.notes.preview.ImagePreviewActivity
import com.elementary.tasks.notes.preview.ImagesSingleton
import com.elementary.tasks.notes.preview.NotePreviewActivity
import com.elementary.tasks.reminder.ReminderBuilderLauncher
import com.elementary.tasks.reminder.preview.adapter.ReminderPreviewDataAdapter
import com.github.naz013.common.intent.IntentKeys
import com.github.naz013.feature.common.livedata.nonNullObserve
import com.github.naz013.logging.Logger
import com.github.naz013.ui.common.activity.BindingActivity
import com.github.naz013.ui.common.activity.toast
import com.github.naz013.ui.common.context.buildIntent
import com.github.naz013.ui.common.context.startActivity
import com.github.naz013.ui.common.login.LoginApi
import com.github.naz013.ui.common.view.applyBottomInsets
import com.github.naz013.ui.common.view.applyTopInsets
import com.github.naz013.usecase.googletasks.TasksIntentKeys
import org.koin.android.ext.android.inject
import org.koin.androidx.viewmodel.ext.android.viewModel
import org.koin.core.parameter.parametersOf
import org.threeten.bp.LocalTime

class ReminderPreviewActivity : BindingActivity<ActivityReminderPreviewBinding>() {

  private val viewModel by viewModel<ReminderPreviewViewModel> { parametersOf(getId()) }
  private val dateTimeManager by inject<DateTimeManager>()
  private val imagesSingleton by inject<ImagesSingleton>()
  private val reminderBuilderLauncher by inject<ReminderBuilderLauncher>()
  private val dialogues by inject<Dialogues>()
  private val prefs by inject<Prefs>()

  private val adapter = ReminderPreviewDataAdapter(
    fragmentManager = supportFragmentManager,
    prefs = prefs,
    onToggleClicked = { viewModel.switchClick() },
    onMapClick = { openFullMap(it) },
    subTaskCheckClick = { viewModel.onSubTaskChecked(it) },
    subTaskRemoveClick = { viewModel.onSubTaskRemoved(it) },
    noteClick = { openNote(it) },
    noteImageClick = { imageId, note -> onNoteImageClicked(imageId, note) },
    googleTaskClick = { onGoogleTaskClicked(it) },
    googleCalendarClick = { openCalendar(it) },
    googleRemoveClick = { viewModel.deleteEvent(it) }
  )

  override fun inflateBinding() = ActivityReminderPreviewBinding.inflate(layoutInflater)

  override fun onCreate(savedInstanceState: Bundle?) {
    enableEdgeToEdge()
    super.onCreate(savedInstanceState)
    binding.dataListView.applyBottomInsets()
    initTopAppBar()
    initViews()
    initViewModel()
  }

  private fun getId() = intentString(IntentKeys.INTENT_ID)

  private fun initViewModel() {
    lifecycle.addObserver(viewModel)
    viewModel.reminderData.nonNullObserve(this) {
      adapter.submitList(it)
      updateMenu()
    }
    viewModel.result.nonNullObserve(this) { commands ->
      when (commands) {
        Commands.DELETED -> closeWindow()
        Commands.FAILED -> toast(getString(R.string.reminder_is_outdated), Toast.LENGTH_SHORT)
        else -> {
        }
      }
    }
    viewModel.sharedFile.nonNullObserve(this) {
      TelephonyUtil.sendFile(this, it)
    }
  }

  private fun openCalendar(id: Long) {
    if (id <= 0L) return
    val uri = Uri.parse("content://com.android.calendar/events/$id")
    val intent = Intent(Intent.ACTION_VIEW, uri)
    runCatching {
      startActivity(intent)
    }
  }

  private fun onGoogleTaskClicked(id: String) {
    LoginApi.openLogged(this, GoogleTaskActivity::class.java) {
      putExtra(IntentKeys.INTENT_ID, id)
      putExtra(TasksIntentKeys.INTENT_ACTION, TasksIntentKeys.EDIT)
    }
  }

  private fun openNote(id: String) {
    startActivity(NotePreviewActivity::class.java) {
      putExtra(IntentKeys.INTENT_ID, id)
    }
  }

  private fun onNoteImageClicked(imageId: Int, note: UiNoteList) {
    val imagePosition = note.images.indexOfFirst { it.id == imageId }.takeIf { it != -1 } ?: 0
    imagesSingleton.setCurrent(
      images = note.images,
      color = note.colorPosition,
      palette = note.colorPalette
    )
    startActivity(ImagePreviewActivity::class.java) {
      addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
      putExtra(IntentKeys.INTENT_ID, note.id)
      putExtra(IntentKeys.INTENT_POSITION, imagePosition)
    }
  }

  private fun shareReminder() {
    viewModel.shareReminder()
  }

  private fun editReminder() {
    reminderBuilderLauncher.openLogged(this) {
      putExtra(IntentKeys.INTENT_ID, getId())
    }
  }

  private fun removeReminder() {
    if (!viewModel.canDelete) {
      dialogues.askConfirmation(this, getString(R.string.move_to_trash)) {
        if (it) viewModel.moveToTrash()
      }
    } else {
      dialogues.askConfirmation(this, getString(R.string.delete)) {
        if (it) viewModel.deleteReminder(true)
      }
    }
  }

  private fun makeCopy() {
    showDialog()
  }

  private fun closeWindow() {
    postUi { finishAfterTransition() }
  }

  private fun showDialog() {
    var time = LocalTime.of(0, 0)
    val list = mutableListOf<LocalTime>()
    val times = mutableListOf<String>()
    var isRunning = true
    do {
      if (time.hour == 23 && time.minute == 30) {
        isRunning = false
      } else {
        list.add(time)
        times.add(dateTimeManager.getTime(time))
        time = time.plusMinutes(30)
      }
    } while (isRunning)
    val builder = dialogues.getMaterialDialog(this)
    builder.setTitle(R.string.choose_time)
    builder.setItems(times.toTypedArray()) { dialog, which ->
      dialog.dismiss()
      saveCopy(list[which])
    }
    builder.create().show()
  }

  private fun saveCopy(time: LocalTime) {
    Logger.d("saveCopy: $time")
    viewModel.copyReminder(time)
  }

  private fun initViews() {
    binding.dataListView.layoutManager = LinearLayoutManager(this)
    binding.dataListView.adapter = adapter
  }

  private fun openFullMap(view: View) {
    val options = ActivityOptions.makeSceneTransitionAnimation(this, view, "map")
    startActivity(
      buildIntent(FullscreenMapActivity::class.java) {
        putExtra(IntentKeys.INTENT_ID, getId())
      },
      options.toBundle()
    )
  }

  private fun initTopAppBar() {
    binding.appBar.applyTopInsets()
    binding.toolbar.setOnMenuItemClickListener { menuItem ->
      return@setOnMenuItemClickListener when (menuItem.itemId) {
        R.id.action_delete -> {
          removeReminder()
          true
        }

        R.id.action_make_copy -> {
          makeCopy()
          true
        }

        R.id.action_share -> {
          shareReminder()
          true
        }

        R.id.action_edit -> {
          editReminder()
          true
        }

        else -> false
      }
    }
    binding.toolbar.setNavigationOnClickListener { closeWindow() }
    updateMenu()
  }

  private fun updateMenu() {
    binding.toolbar.menu.also { menu ->
      menu.getItem(2)?.isVisible = viewModel.canCopy
    }
  }
}
